File: Python/110.py
# Initial thoughts
    # A height-balanced binary tree is a binary tree in which the depth of the 
        # two subtrees of every node never differs by more than one
    # Note: just compiling leaf depths or first-null-depths will not work for
        # this because of test cases like [1,2,3,4,5,6,null,8]
    # Proper idea: check for equality between depths of left/right subtree
        # and recursively apply to the subtrees
# Improved algorithm
    # To avoid dfs recalculations, build solution by bottom up by finding
        # height from 1 + max(l, r) instead of doing DFS at each node
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(root):
            if not root: return [True, 0]
            left = dfs(root.left)
            right = dfs(root.right)
            balance = abs(left[1] - right[1]) <= 1 \
            and left[0] and right[0]
            return [balance, 1 + max(left[1], right[1])]

        return dfs(root)[0]

File: C++/110.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int height = 0;
        return dfs(root, height);
    }
private:
    bool dfs(TreeNode* root, int& height){
        if(root == NULL){
            height = -1;
            return true;
        }
        int left = 0;
        int right = 0;
        if(!dfs(root -> left, left) || !dfs(root -> right, right)) return false;
        if(abs(left - right) > 1) return false;
        height = 1 + max(left, right);
        return true;
    }
};
/*
- Idea: use DFS to determine height of each node recursively
    - at each call, stipulate that height distance must be less than 2
- Note: first approach failed because you have to check that subtrees are balanced at each
    - step instead of just at the end
*/